### DISCUSSION TITLE
Installer for Windows

### DISCUSSION THREAD

**Discussion by icedterminal** [liquidctl/liquidctl#375] [Ideas]
Wondering if you'd be interested in packaging the EXE into an installer that can be signed. Would this help with false-positives? I have an example here

https://github.com/icedterminal/liquidctl_installer
https://github.com/icedterminal/liquidctl_installer/releases/tag/1.0.0

This would greatly improve the user experience.

**Comment by jonasmalacofilho** [comment]
The installer by itself would already improve user experience.  It would also allow us to distribute liquidctl through winget (#287), which is something I'm excited about since it may finally bring the install-and-(safely)-forget experience we're accustomed to on Linux/Mac to Windows.

Code signing really is orthogonal to all of that: in theory we could simply code sign the executable in our current zip file.  But it should indeed help with AV false-positives, even if only partially (maybe more so with the "extended validation (EV)" certificate).  Still, it's expensive for what it is (250~500 USD/year for the standard/cheaper certificate), plus it seems that it can sometimes be a bit of a hassle.

**Comment by icedterminal** [comment]
https://github.com/icedterminal/liquidctl_installer/releases/tag/1.0.1

This further improves the installer.

- Disable modify since there is nothing to actually modify in regards to installing.
- Change build to x64. (still built as debug)
- Change banner color to white.
- Correctly display EULA and Acknowledgements.

I have tested upgrades/updates and they work as intended. Future updates will apply seamlessly. Uninstalls cleanly. Downgrades prohibited unless you uninstall first. This is a recommended practice of installation development. As of now, this installs per-machine requiring admin rights. Is this needed to run any of the components within? I cannot test this since I no longer have any devices that are supported. I have a custom HydroX loop by Corsair.

**Reply by jonasmalacofilho** [reply]
> As of now, this installs per-machine requiring admin rights. Is this needed to run any of the components within?

Admin rights should not be necessary to use `liquidctl` on Windows.

**Reply by icedterminal** [reply]
I've looked over the implications for either way. It's your call on how you want to handle it. Indeed liquidctl does not require any admin rights. In the off chance that someone needs to make adjustment and they have multiple users, they won't need to install multiple copies of liquidctl per user. Just open a Terminal windows in user mode and make changes. I think having it install per-machine is beneficial.

Further, could you incorporate a version number into the exe? This way, during the installer building, it can pull the version number from the exe entirely automating the process. You won't need to open the solution to bump the version there too.

**Reply by jonasmalacofilho** [reply]
> I think having it install per-machine is beneficial.

I haven't given this much thought, but I tend to agree with you.

> Further, could you incorporate a version number into the exe?

isn't `liquidctl --version` sufficient?

**Reply by icedterminal** [reply]
No. Not in the way the installer build works. Right now, you have to type this in manually each time you release a new version.

`<Product [...] Version="1.7.1" [...]>` ([reference](https://github.com/icedterminal/liquidctl_installer/blob/0c8134d8935691489721d937ee7bba42d5c940a8/liquidctl_installer/Code/Product.wxs#L12))

This version is what gets shown under Add and Remove, as well what triggers the upgrade procedure. A new version gives a product GUID on build.

Because the EXE does not include a "Product" or "File" version.:

![image](https://user-images.githubusercontent.com/4755734/132966273-de32d55a-d08d-4ef8-914b-993b15ae46f7.png)

the build process has no metadata to pull from the EXE and update it programmatically. This would streamline the build process. One script to build the program itself and then rebuild the installer. Otherwise, once liquidctl is built, you have to open the solution, bump the version and then build.

```
<your script stuff to build liquidctl here>
msbuild liquidctl_installer.sln -t:rebuild
```

edit: My intent is to entirely replace all versioning in the project files with a variable that grabs the value found in the EXE. Thus no need for extra commands or manually opening files.
